include/qdmi/device/types.h

```diff
@@ -0,0 +1,58 @@
+/*------------------------------------------------------------------------------
+Part of the MQSS Project, under the Apache License v2.0 with LLVM Exceptions.
+See https://llvm.org/LICENSE.txt for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+------------------------------------------------------------------------------*/
+
+/** @file
+ * @brief Defines all types used within QDMI.
+ */
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// The following disables the clang-tidy warning modernize-use-using.
+// Since this is C code, we cannot use the using keyword.
+// NOLINTBEGIN(modernize-use-using)
+
+/**
+ * @brief Opaque type for a job.
+ * @details A job is a handle to manage the job after it is submitted and to
+ * access the results.
+ *
+ * @par
+ * This type is implemented by the driver.
```

Comment:
Is `QDMI_Job` implemented by a driver or a device?

What is the reasoning behind moving it from `qdmi/common` to `qdmi/device` since it is also used by QDMI client?
If it is because it is implemented by a device, should QDMI_device and QDMI_session also be moved to `qdmi/driver` as they are implemented by a driver?

Reply:
`QDMI_Job` is implemented by each device. The main reason for moving these device-defined types to their own header is to make the prefixing possible/feasible. Each device will define its own (prefixed) version of these types. On the other hand, the (enum) types in `common/types.h` are not intended to be touched by device implementers (e.g., they should not be adding new enums or enum values).
I agree though, that it could make sense to move the `QDMI_Session` and `QDMI_device` to a `driver/types.h` file.
@ystade what do you think?

Reply:
To note that first, @mnfarooqi has definitely a point in that this is wrong in the documentation and should be changed before merging this.

The motivation for the `device/types.h` was to be able to copy the `device.h` and `device/*` in order to have everything that must be prefixed for the respective device implementation.

`QDMI_Session` and `QDMI_device` culd definitely be moved to the `driver`. Then I think the responsibilities should be clear.

Reply:
> QDMI_Session and QDMI_device could definitely be moved to the driver. Then I think the responsibilities should be clear.

Done in https://github.com/Munich-Quantum-Software-Stack/QDMI/pull/86/commits/62ce706294645bc305b6d03febcd0ea7e0766e8a

Reply:
> To note that first, @mnfarooqi has definitely a point in that this is wrong in the documentation and should be changed before merging this.

Done in https://github.com/Munich-Quantum-Software-Stack/QDMI/pull/86/commits/8c3ed5e427a155037f8ebf7dd4fa20189e3b4ccd